### PR TITLE
Feature: shared ownership

### PR DESCRIPTION
**Pull request by cap10morgan**
Implements #31.

Adds new `team (add | list | remove)` and `whoami` commands. Asks for username on init and stores it in git config. A global default username can be configured with `git config --global dgit.username [username]`, but it also looks for any configured GitHub username and suggests that for your dgit username. It then suggests `username/basedir` for the full repo name.

It can also migrate the old `dgit.default` keyring entries to the new `dgit.[username]` entries.

And a few other small fixes & improvements (which I will pull into their own PR if controversial).

Tests shmests... ;)

### CODE REVIEW

**Review by brandonwestcott — APPROVED** [review]
A few questions / comments, but lgtm 👍 

It seems the key migration is handled automatically? Are there any other effects or manual steps for existing users?

**Comment by brandonwestcott** [comment]
Also had a thought on this over the weekend, but needs discussion and certainly can be a future improvement:

What if we stored the team membership in files in the repo, then on push to the HEAD branch (still needs implemented) we updated the tupelo authentications array? Then any future branch-limiting conditions, such as master can only be updated after # thumbs on a PR, will benefit from that protection.

**Comment by cap10morgan** [comment]
> A few questions / comments, but lgtm 👍
> 
> It seems the key migration is handled automatically? Are there any other effects or manual steps for existing users?

Yes it should be. The other effect is that if the user sets a dgit username that is different than what's in their `~/.gitconfig` file, it will be written to the local repo's `./.git/config` file.

File: initializer/initializer.go [hunk 1]
```diff
@@ -103,41 +174,62 @@ func (i *Initializer) getAuth() (transport.AuthMethod, error) {
...
 	if err != nil {
 		return nil, fmt.Errorf("Error fetching key from keyring: %v", err)
 	}
 
+	i.auth = dgit.NewPrivateKeyAuth(privateKey)
+
 	if isNew {
+		ctx := context.Background()
+
+		opts := &usertree.Options{
```
[the remark this answers is not in the document]
Comment (cap10morgan) [reply initializer/initializer.go:194]: Yeah probably. I can try it out.

File: initializer/initializer.go [hunk 2]
```diff
@@ -84,14 +99,70 @@ func (i *Initializer) Init(ctx context.Context, args []string) error {
...
+	dgitConfig := repoConfig.Raw.Section(constants.DgitConfigSection)
+
+	var username string
+	if dgitConfig != nil {
+		username = dgitConfig.Option("username")
+	}
+
+	// try looking up the GitHub username for the default
+	if username == "" {
+		if ghConfig := repoConfig.Raw.Section("github"); ghConfig != nil {
```
[the remark this answers is not in the document]
Comment (cap10morgan) [reply initializer/initializer.go]: It is! Various git tools I use look for it (hub, Tower, etc.).

File: transport/dgit/client.go [hunk 3]
```diff
@@ -50,44 +50,133 @@ func Default() (*Client, error) {
...
 	return server.NewServer(loader).NewUploadPackSession(ep, auth)
 }
 
 func (c *Client) NewReceivePackSession(ep *transport.Endpoint, auth transport.AuthMethod) (transport.ReceivePackSession, error) {
-	loader := NewChainTreeLoader(c.ctx, c.tupelo, c.nodestore, auth)
+	loader := NewChainTreeLoader(c.ctx, c.Tupelo, c.Nodestore, auth)
 	return server.NewServer(loader).NewReceivePackSession(ep, auth)
 }
+
+func (c *Client) AddRepoCollaborator(ctx context.Context, repo *Repo, collaborators []string) error {
```
[the remark this answers is not in the document]
Comment (cap10morgan) [reply transport/dgit/client.go:103]: I can try this out too and see how it compares.

File: initializer/initializer.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (cap10morgan) [reply initializer/initializer.go:194]: Done in e5d7671.

File: transport/dgit/client.go [hunk 4]
```diff
@@ -50,44 +49,131 @@ func Default() (*Client, error) {
 func NewClient(ctx context.Context, basePath string) (*Client, error) {
 	var err error
 	c := &Client{ctx: ctx}
-	dir := path.Join(basePath, protocol)
-	c.tupelo, c.nodestore, err = clientbuilder.Build(ctx, dir)
+	dir := path.Join(basePath, constants.Protocol)
+	c.Tupelo, c.Nodestore, err = clientbuilder.Build(ctx, dir)
```
Comment (brandonwestcott) [inline transport/dgit/client.go:53]: @cap10morgan just to note related to this, I added `Client.DagStore()` method in https://github.com/quorumcontrol/tupelo-go-sdk/pull/212/files#diff-7f6280b5f587f41a183254e1b182f3c9R100 so we don't have to keep passing around both vars everywhere

**Review by brandonwestcott — APPROVED** [review]

File: transport/dgit/client.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (cap10morgan) [reply transport/dgit/client.go:100]: Looked into this quickly, and I think there definitely could be something to it. The conceptual boundaries of the new `Repo` stuff is still in flux for sure. But I think I'd want to save it for a potential future refactor at this point.

**Pull request merged**